source/opt/common_uniform_elim_pass.h

```diff
@@ -0,0 +1,214 @@
+
+  // Returns the id of the merge block declared by a merge instruction in 
+  // this block, if any.  If none, returns zero.
+  uint32_t MergeBlockIdIfAny(const ir::BasicBlock& blk, uint32_t* cbid);
+
+  // Compute structured successors for function |func|.
+  // A block's structured successors are the blocks it branches to
+  // together with its declared merge block if it has one.
+  // When order matters, the merge block always appears first.
+  // This assures correct depth first search in the presence of early 
+  // returns and kills. If the successor vector contain duplicates
+  // if the merge block, they are safely ignored by DFS.
+  void ComputeStructuredSuccessors(ir::Function* func);
+
+  // Compute structured block order for |func| into |structuredOrder|. This
+  // order has the property that dominators come before all blocks they
+  // dominate and merge blocks come after all blocks that are in the control
+  // constructs of their header.
+  void ComputeStructuredOrder(ir::Function* func,
+      std::list<ir::BasicBlock*>* order);
+
+  // Eliminate loads of uniform variables which have previously been loaded.
+  // If first load is in control flow, move it to first block of function.
+  // Most effective if preceded by UniformAccessChainRemoval().
+  bool CommonUniformLoadElimination(ir::Function* func);
+
+  // Eliminate loads of uniform sampler and image variables which have previously
+  // been loaded in the same block for types whose loads cannot cross blocks.
+  bool CommonUniformLoadElimBlock(ir::Function* func);
+
+  // Eliminate duplicated extracts of same id. Extract may be moved to same
+  // block as the id definition. This is primarily intended for extracts
+  // from uniform loads. Most effective if preceded by
+  // CommonUniformLoadElimination().
+  bool CommonExtractElimination(ir::Function* func);
+
+  // Initialize the label2SSA map entry for a block. Insert phi instructions
+  // into block when necessary. All predecessor blocks must have been
+  // For function |func|, first change all constant index
+  // access chain loads into equivalent composite extracts. Then consolidate 
+  // identical uniform loads into one uniform load. Finally, consolidate
+  // identical uniform extracts into one uniform extract. This may require
+  // moving a load or extract to a point which dominates all uses.
+  // Return true if func is modified. 
+  //
+  // This pass requires the function to have structured control flow ie shader
+  // capability. It also requires logical addressing ie Addresses capability
+  // is not enabled. It also currently does not support any extensions.
+  bool EliminateCommonUniform(ir::Function* func);
+
+  // Initialize extensions whitelist
+  void InitExtensions();
+
+  // Return true if all extensions in this module are allowed by this pass.
+  bool AllExtensionsSupported() const;
+
+  // Return true if all uses of varId are only through supported reference
```

Comment:
Re OpMemberDecorate, in this context I was only concerned with Decorates that could be applied to non-type instructions, so I have renamed this function to IsNonTypeDecorate, and fixed the comment.

As for OpGroupDecorate, I would rather not redundantly document this in optimizer.hpp for each pass which does not support it, which are numerous. I do have TODOs in each pass file that prohibits OpGroupDecorate. I have also opened an issue so that users can see this more readily. Will that suffice?